File: ChessLab/Laboratory/LabStartPosition.swift
import ChessKit
import Foundation

/// Résout une position de départ de série Laboratoire à partir d'un texte qui
/// peut être un **FEN** OU un **PGN**. Un PGN est ramené à la FEN de sa
/// position FINALE (ligne principale) — pour lancer une série depuis la fin
/// d'une ouverture collée/importée.
enum LabStartPosition {

    /// - Returns: la FEN de départ, le nombre de demi-coups joués pour
    ///   l'atteindre (`0` pour un FEN direct), et si la source était un PGN.
    ///   `nil` si le texte n'est ni un FEN légal ni un PGN lisible.
    static func resolve(_ text: String) -> (fen: String, plies: Int, fromPGN: Bool)? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        // Un FEN légal l'emporte : on le prend tel quel.
        if FENValidator.isLegal(trimmed) {
            return (trimmed, 0, false)
        }

        // Sinon, on tente le PGN (première partie d'un éventuel multi-parties),
        // et on remonte sa position finale.
        let candidate = PGNSanitizer.sanitize(PGNSanitizer.splitIntoGames(trimmed).first ?? trimmed)
        // PGNLoader : même robustesse que la bibliothèque et l'analyse — un
        // PGN avec prise en passant doit pouvoir donner sa position finale.
        guard !candidate.isEmpty, let game = PGNLoader.game(from: candidate) else { return nil }

        var index = game.startingIndex
        var plies = 0
        while game.moves.hasIndex(after: index) {
            index = game.moves.index(after: index)
            plies += 1
        }
        guard let fen = game.positions[index]?.fen else { return nil }
        return (fen, plies, true)
    }
}

File: android/app/src/main/java/com/chesslab/lab/LabStartPosition.kt
package com.chesslab.lab

import chesskit.MoveTree
import chesskit.PgnParser
import chesskit.Position
import com.chesslab.analysis.PgnSanitizer
import com.chesslab.courses.CourseRepository

/**
 * Résout la position de départ d'une série à partir d'un texte qui peut être
 * un **FEN** ou un **PGN**. Pendant de `LabStartPosition.swift`.
 *
 * Un PGN est ramené à la FEN de sa position FINALE (ligne principale) : c'est
 * ce qui permet de lancer une série depuis la fin d'une ouverture qu'on vient
 * de coller, sans avoir à en recopier la position à la main.
 */
object LabStartPosition {

    /**
     * @return la FEN de départ, le nombre de demi-coups joués pour l'atteindre
     * (`0` pour un FEN direct), et si la source était un PGN. `null` si le
     * texte n'est ni un FEN légal ni un PGN lisible.
     */
    fun resolve(text: String): Resolved? {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return null

        // Un FEN légal l'emporte : on le prend tel quel. `CourseRepository`
        // plutôt que le parseur brut, pour accepter les quatre champs — et
        // pour refuser une chaîne qui n'a pas ses deux rois.
        CourseRepository.position(trimmed)?.let {
            return Resolved(it.fen, plies = 0, fromPgn = false)
        }

        val candidate = PgnSanitizer.splitIntoGames(PgnSanitizer.sanitize(trimmed))
            .firstOrNull() ?: trimmed
        if (candidate.isEmpty()) return null
        val game = runCatching { PgnParser.parse(candidate) }.getOrNull() ?: return null

        val mainline = game.moves.indices
            .filter { it.variation == MoveTree.Index.MAIN_VARIATION }
            .sorted()
        val last = mainline.lastOrNull() ?: return null
        val position = game.position(last) ?: return null
        return Resolved(position.fen, plies = mainline.size, fromPgn = true)
    }

    data class Resolved(val fen: String, val plies: Int, val fromPgn: Boolean) {
        val isStandard: Boolean get() = fen == Position.standard.fen
    }
}
